### ISSUE TITLE
xcodebuild: error: Scheme Transient Testing

### ISSUE THREAD

**Issue by hactar** [lyndsey-ferguson/fastlane-plugin-test_center#142]
### New Issue Checklist

- [x] Updated `fastlane-plugin-test_center` to the latest version
- [x] I read the [README.md](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/blob/master/README.md)
- [x] I reviewed the [example(s)](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/blob/master/README.md) for the action(s) I am using
- [x] I have removed any sensitive data such as passwords, authentication tokens, or anything else I do not want to world to see

### Issue Description
We've just updated to the latest Xcode GM (Seed 2) and latest fastlane, and test center has stopped working. When running, it now exits with the log below and an error of ```undefined method `size' for nil:NilClass``` because xcodebuild fails with the error `xcodebuild: error: Scheme Transient Testing is not currently configured for the build-for-testing action.`

fastlane scan works, and removing the `build-for-testing` argument from the xcodebuild command also allows the build to run (not sure how to get sweep to remove that argument, or if that would work at all)

<details>
  <pre>[00:31:51]: Starting test run 1
[00:31:51]: Starting scan #1 with 8 tests for batch #1.

+------------------------------------------------+-------------------------------------------------------------------+
|                                              Summary for scan 2.131.0                                              |
+------------------------------------------------+-------------------------------------------------------------------+
| workspace                                      | /Users/patrick/Documents/localgit/APPNAME-x-ios/APPNAME.xcworksp  |
|                                                | ace                                                               |
| fail_build                                     | true                                                              |
| scheme                                         | APPNAME                                                           |
| skip_detect_devices                            | false                                                             |
| force_quit_simulator                           | false                                                             |
| reset_simulator                                | false                                                             |
| reinstall_app                                  | false                                                             |
| app_identifier                                 | com.APPNAME                                            |
| only_testing                                   | ["APPNAMETests/BMQuadTreeTests/testLargeTreeNearestSearch",       |
|                                                | "APPNAMETests/BMQuadTreeTests/testLargeTreeSetup",                |
|                                                | "APPNAMETests/BMQuadTreeTests/testMTKDuplicates",                 |
|                                                | "APPNAMETests/BMDirectionsTest/testDirectionResponse",            |
|                                                | "APPNAMETests/BMNavigationSETests/testBarcelona",                 |
|                                                | "APPNAMETests/BMNavigationSETests/testDegreeMod",                 |
|                                                | "APPNAMETests/BMNavigationSETests/testSaikogasse",                |
|                                                | "APPNAMETests/BMNavigationSETests/testABCD"]                      |
| xctestrun                                      | build/DerivedData/Build/Products/APPNAME_iphonesimulator13.0-x86  |
|                                                | _64.xctestrun                                                     |
| clean                                          | false                                                             |
| open_report                                    | false                                                             |
| output_directory                               | /Users/patrick/Documents/localgit/APPNAME-x-ios/fastlane/test_ou  |
|                                                | tput/APPNAMETests-batch-1                                         |
| output_files                                   | report.junit                                                      |
| buildlog_path                                  | ~/Library/Logs/scan                                               |
| include_simulator_logs                         | false                                                             |
| derived_data_path                              | build/DerivedData                                                 |
| should_zip_build_products                      | false                                                             |
| result_bundle                                  | false                                                             |
| use_clang_report_name                          | false                                                             |
| disable_concurrent_testing                     | true                                                              |
| skip_build                                     | false                                                             |
| xcargs                                         | -UseNewBuildSystem=YES build-for-testing                          |
|                                                | -parallel-testing-enabled NO                                      |
| slack_use_webhook_configured_username_and_icon | false                                                             |
| slack_username                                 | fastlane                                                          |
| slack_icon_url                                 | https://s3-eu-west-1.amazonaws.com/fastlane.tools/fastlane.png    |
| skip_slack                                     | false                                                             |
| slack_only_on_failure                          | false                                                             |
| xcodebuild_command                             | env NSUnbufferedIO=YES xcodebuild                                 |
| output_types                                   | junit                                                             |
| xcode_path                                     | /Applications/Xcode.app                                           |
| devices                                        | ["iPhone 8"]                                                      |
+------------------------------------------------+-------------------------------------------------------------------+

[00:31:51]: $ set -o pipefail && env NSUnbufferedIO=YES xcodebuild -destination 'platform=iOS Simulator,id=C4761312-6DA0-4DB7-871A-6159B83FA922' -derivedDataPath 'build/DerivedData' -disable-concurrent-testing -xctestrun 'build/DerivedData/Build/Products/APPNAME_iphonesimulator13.0-x86_64.xctestrun' -UseNewBuildSystem=YES build-for-testing -parallel-testing-enabled NO -only-testing:APPNAMETests/BMQuadTreeTests/testLargeTreeNearestSearch -only-testing:APPNAMETests/BMQuadTreeTests/testLargeTreeSetup -only-testing:APPNAMETests/BMQuadTreeTests/testMTKDuplicates -only-testing:APPNAMETests/BMDirectionsTest/testDirectionResponse -only-testing:APPNAMETests/BMNavigationSETests/testBarcelona -only-testing:APPNAMETests/BMNavigationSETests/testDegreeMod -only-testing:APPNAMETests/BMNavigationSETests/testSaikogasse -only-testing:APPNAMETests/BMNavigationSETests/testABCD test-without-building | tee '/Users/patrick/Library/Logs/scan/APPNAME-APPNAME.log' | xcpretty  --report junit --output '/Users/patrick/Documents/localgit/APPNAME-x-ios/fastlane/test_output/APPNAMETests-batch-1/report.junit' --report junit --output '/var/folders/wx/pd_ryg4x0dx1q3nyt08gcylc0000gq/T/junit_report20190918-32016-ihoqch'
[00:31:51]: ▸ Loading...
[00:31:53]: ▸ xcodebuild: error: Scheme Transient Testing is not currently configured for the build-for-testing action.
xcodebuild: error: Scheme Transient Testing is not currently configured for the build-for-testing action.
[00:31:53]: Exit status: 66
[00:31:53]:
+------------------+-----------+
|         Lane Context         |
+------------------+-----------+
| DEFAULT_PLATFORM | ios       |
| PLATFORM_NAME    | ios       |
| LANE_NAME        | ios sweep |
+------------------+-----------+
[00:31:53]: Error in your Fastfile at line 102
[00:31:53]:     100:	lane :sweep do
[00:31:53]:     101:	  test_run_block = lambda do |testrun_info|
[00:31:53]:  => 102:	    failed_test_count = testrun_info[:failed].size
[00:31:53]:     103:
[00:31:53]:     104:	    if failed_test_count > 0

+------+----------------------------+-------------+
|                fastlane summary                 |
+------+----------------------------+-------------+
| Step | Action                     | Time (in s) |
+------+----------------------------+-------------+
| 1    | Verifying fastlane version | 0           |
| 2    | default_platform           | 0           |
| 💥   | multi_scan                 | 209         |
+------+----------------------------+-------------+

[00:31:53]: fastlane finished with errors</pre>
</details>

### Environment
(Reduced for privacy, if anything is missing that you need, please comment - but the invocation  of test center is copy pasted from the readme.)

<details>
  <pre><details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                       |
| --------------------------- | ------------------------------------------- |
| OS                          | 10.15                                       |
| Ruby                        | 2.6.3                                       |
| Bundler?                    | true                                        |
| Git                         | git version 2.21.0 (Apple Git-122)          |
| Installation Source         | ~/.rbenv/versions/2.6.3/bin/fastlane        |
| Host                        | Mac OS X 10.15 (19A558d)                    |
| Ruby Lib Dir                | ~/.rbenv/versions/2.6.3/lib                 |
| OpenSSL Version             | OpenSSL 1.0.2s  28 May 2019                 |
| Is contained                | false                                       |
| Is homebrew                 | false                                       |
| Is installed via Fabric.app | false                                       |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/ |
| Xcode Version               | 11.0                                        |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   | en_US.UTF-8 | ✅ |
| LANGUAGE |             |   |

</details>

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.131.0 | ✅ Up-To-Date  |

### Loaded fastlane plugins:

| Plugin                      | Version | Update-Status |
| --------------------------- | ------- | ------------- |
| fastlane-plugin-test_center | 3.8.3   | ✅ Up-To-Date  |

<details><summary><b>Loaded gems</b></summary>

| Gem                         | Version      |
| --------------------------- | ------------ |
| did_you_mean                | 1.3.0        |
| bundler                     | 1.17.2       |
| etc                         | 1.0.1        |
| forwardable                 | 1.2.0        |
| CFPropertyList              | 3.0.1        |
| public_suffix               | 2.0.5        |
| addressable                 | 2.7.0        |
| atomos                      | 0.1.3        |
| babosa                      | 1.0.2        |
| claide                      | 1.0.3        |
| colored                     | 1.2          |
| colored2                    | 3.1.2        |
| colorize                    | 0.8.1        |
| highline                    | 1.7.10       |
| commander-fastlane          | 4.4.6        |
| declarative                 | 0.0.10       |
| declarative-option          | 0.1.0        |
| digest-crc                  | 0.4.1        |
| unf_ext                     | 0.0.7.6      |
| unf                         | 0.1.4        |
| domain_name                 | 0.5.20190701 |
| dotenv                      | 2.7.5        |
| emoji_regex                 | 1.0.1        |
| excon                       | 0.66.0       |
| multipart-post              | 2.0.0        |
| faraday                     | 0.15.4       |
| http-cookie                 | 1.0.3        |
| faraday-cookie_jar          | 0.0.6        |
| faraday_middleware          | 0.13.1       |
| fastimage                   | 2.1.7        |
| gh_inspector                | 1.1.3        |
| jwt                         | 2.1.0        |
| memoist                     | 0.16.0       |
| multi_json                  | 1.13.1       |
| os                          | 1.0.1        |
| signet                      | 0.11.0       |
| googleauth                  | 0.6.7        |
| httpclient                  | 2.8.3        |
| mime-types-data             | 3.2019.0904  |
| mime-types                  | 3.3          |
| uber                        | 0.1.0        |
| representable               | 3.0.4        |
| retriable                   | 3.1.2        |
| google-api-client           | 0.23.9       |
| google-cloud-env            | 1.2.1        |
| google-cloud-core           | 1.3.1        |
| google-cloud-storage        | 1.16.0       |
| json                        | 2.2.0        |
| mini_magick                 | 4.9.5        |
| multi_xml                   | 0.6.0        |
| plist                       | 3.5.0        |
| rubyzip                     | 1.2.4        |
| security                    | 0.1.3        |
| naturally                   | 2.2.0        |
| simctl                      | 1.6.6        |
| slack-notifier              | 2.3.2        |
| terminal-notifier           | 2.0.0        |
| unicode-display_width       | 1.6.0        |
| terminal-table              | 1.8.0        |
| tty-screen                  | 0.7.0        |
| tty-cursor                  | 0.7.0        |
| tty-spinner                 | 0.9.1        |
| word_wrap                   | 1.0.0        |
| nanaimo                     | 0.2.6        |
| xcodeproj                   | 1.12.0       |
| rouge                       | 2.0.7        |
| xcpretty                    | 0.3.0        |
| xcpretty-travis-formatter   | 1.0.0        |
| xctest_list                 | 1.1.8        |
| fastlane-plugin-test_center | 3.8.3        |
</details>

*generated on:* **2019-09-18**
</details></pre>
</details>

**Comment by lyndsey-ferguson** [comment]
@hactar thanks for the bug report. I want to make everything explicit, so I have some questions.

1. You write that `scan` works. Are you saying that if you pass [`:build_for_testing: true`](https://github.com/fastlane/fastlane/blob/master/scan/lib/scan/options.rb#L280) to `scan`, this failure does not happen?

2. You write that if you remove that command from `xcodebuild`, it does not fail also? Are you stating that you have run the generated `xcodebuild` command directly and it fails in the same way? 

3. I looked at the log that you posted, and it is telling `xcodebuild` to both `build-for-testing` and `test-without-building`. That could be a problem, either `multi_scan` is not properly removing that, or perhaps `scan` is adding that.

```
xcodebuild -destination 'platform=iOS Simulator,id=C4761312-6DA0-4DB7-871A-6159B83FA922' -derivedDataPath 'build/DerivedData' -disable-concurrent-testing -xctestrun 'build/DerivedData/Build/Products/APPNAME_iphonesimulator13.0-x86_64.xctestrun' -UseNewBuildSystem=YES build-for-testing -parallel-testing-enabled NO -only-testing:APPNAMETests/BMQuadTreeTests/testLargeTreeNearestSearch -only-testing:APPNAMETests/BMQuadTreeTests/testLargeTreeSetup -only-testing:APPNAMETests/BMQuadTreeTests/testMTKDuplicates -only-testing:APPNAMETests/BMDirectionsTest/testDirectionResponse -only-testing:APPNAMETests/BMNavigationSETests/testBarcelona -only-testing:APPNAMETests/BMNavigationSETests/testDegreeMod -only-testing:APPNAMETests/BMNavigationSETests/testSaikogasse -only-testing:APPNAMETests/BMNavigationSETests/testABCD test-without-building | tee '/Users/patrick/Library/Logs/scan/APPNAME-APPNAME.log' | xcpretty  --report junit --output '/Users/patrick/Documents/localgit/APPNAME-x-ios/fastlane/test_output/APPNAMETests-batch-1/report.junit' --report junit --output '/var/folders/wx/pd_ryg4x0dx1q3nyt08gcylc0000gq/T/junit_report20190918-32016-ihoqch'
```

**Comment by lyndsey-ferguson** [comment]
Oh, one more question. If you run `multi_scan` with the non-beta of Xcode, do you see `build-for-testing` in `xcodebuild` when running the tests?

**Comment by hactar** [comment]
1. Yes, if I run bundle exec fastlane scan the build_for_testing is included and it works
2. If I take the command that fastlane prints just before failing and remove the build_for_testing flag, it works too, though I'm guessing this is not the solution. I can also remove the -xctestrun parameter and replace it with -scheme (like the bundle exec fastlane scan does), the issue goes away too, I'm guessing something in xctestrun is the culprit?
3. Adding or removing `test-without-building` has no effect.

I currently have older version of xcode around - if you need me to install it ping again and I'll see what it throughs... (as far as i remember build-for-testing was always there)

**Comment by lyndsey-ferguson** [comment]
So, it seems from your report that there is no change in the failure in the `xcodebuild` command when `test-without-building` and `build-for-testing` are both in the command. I don't have a lot of time to debug this, so this Issue will have to wait _unless_ you are able to help.

If you can help, I can make it simple:
1. In your `Gemfile`, make sure that you include the `pry-byebug` gem: `gem 'pry-byebug'`
2. `bundle install` to install the `pry-byebug` gem
3. In your `Fastfile`, at the top of the file, add `require 'pry-byebug`, and before the call to `multi_scan`, add on a new line `byebug`
4. Run your fastlane
5. When it breaks, type `b TestCenter::Helper::MultiScanManager::RetryingScan.#update_scan_options` to create a new breakpoint in the class.
6. Type `c` to continue
7. When your code breaks at that method, type `n` 3 times to step over the code until the `scan_options` are assigned.
8. Type 'puts scan_options` and copy and paste the output to this issue.

**Comment by lyndsey-ferguson** [comment]
Okay, so I see this:
```
[00:31:53]:  => 102:	    failed_test_count = testrun_info[:failed].size
```
It is failing in your `:testrun_completed_block` callback. If you _do not_ pass in that as an option, does it still crash? It should not.

I see that no tests were run, that would explain the inability of `multi_scan` to read a junit file of tests because it was not created.

Since my steps that I provided did not provide useful information, I'll create a branch that can log that out.

**Comment by lyndsey-ferguson** [comment]
Aha, and, now that I have had time to look more thoroughly at your provided data, I see that you are asking `xcodebuild` to build-for-testing when running the tests!
```
| xcargs                                         | -UseNewBuildSystem=YES build-for-testing                          |
```
I can add code to warn of that issue.

**Comment by lyndsey-ferguson** [comment]
Can you modify your `Pluginfile` per my instructions below, run `bundle install`, and then run your fastlane again?

Pluginfile:
```
gem 'fastlane-plugin-test_center', :git => "https://github.com/lyndsey-ferguson/fastlane-plugin-test_center.git", :branch => "issue-142-scheme-transitent-testing-crash"
```

If the crash continues, please attach the console output as a text file to this issue (makes it easier for me to review). If you do not experience a crash, please let me know.

**Comment by OlKir** [comment]
Thank you very much for the prompt response!!
Unfortunately crash still happens even if I don't pass testrun_completed_block and use branch which you provided for the plugin source.
It seems that xcodebuild crashes itself. But now I see the reason:
```
[11:47:06]: ▸ xcodebuild: error: Scheme Transient Testing is not currently configured for the build-for-testing action.
xcodebuild: error: Scheme Transient Testing is not currently configured for the build-for-testing action.
[11:47:06]: Exit status: 66
```
I checked my scheme and it looks fine on the first look, test bundles are added correctly.
But scheme name is missing in the xcodebuild parameters - could it be the case?
```
set -o pipefail && env NSUnbufferedIO=YES xcodebuild -destination 'platform=iOS Simulator,id=AD30C630-9288-44FB-B96B-99F283B6BF95' -derivedDataPath 'build/DerivedData' -disable-concurrent-testing -xctestrun 'build/DerivedData/Build/Products/Bikemap_iphonesimulator13.0-x86_64.xctestrun' -UseNewBuildSystem=YES build-for-testing -parallel-testing-enabled NO -only-testing:BikemapTests/BMQuadTreeTests/testLargeTreeNearestSearch -only-testing:BikemapTests/BMQuadTreeTests/testLargeTreeSetup -only-testing:BikemapTests/BMQuadTreeTests/testMTKDuplicates -only-testing:BikemapTests/BMDirectionsTest/testDirectionResponse -only-testing:BikemapTests/BMNavigationSETests/testBarcelona -only-testing:BikemapTests/BMNavigationSETests/testDegreeMod -only-testing:BikemapTests/BMNavigationSETests/testSaikogasse -only-testing:BikemapTests/BMNavigationSETests/testABCD test-without-building | tee '/Users/oleksii/Library/Logs/scan/Bikemap-Bikemap.log' | xcpretty  --report junit --output '/Data/Bikemap/bikemap-x-ios/fastlane/test_output/BikemapTests-batch-1/report.junit' --report junit --output '/var/folders/rj/69jcm5w12sn3288smsn20t880000gn/T/junit_report20190925-44619-1lx4sa'
```
May be previous version of xcode took proper one by default and current one tries to use some intermediate Transient Testing (I don't have scheme with such name in the project) and doesn't set it properly?

**Comment by lyndsey-ferguson** [comment]
Is your Scheme named `Transient Testing`? Have you searched the internet for this particular error? I see a [stackoverflow question answered](https://stackoverflow.com/questions/30481630/scheme-is-not-configured-for-the-test-action-ios-xcode-project) on this issue and perhaps that helps?

**Comment by OlKir** [comment]
Nope, my scheme is named differently (Bikemap).
I have seen this stackowerflow question and answers, and have checked if it covers my case. But unfortunately it appeared to be general error message, and doesn't point to the real cause.

When I tried to run xcodebuild with same parameters and just added scheme manually I got an error
`xcodebuild: error: Cannot use -xctestrun with -project, -workspace or -scheme options.`
So now I suspect that incompatible parameters are mixed in the same command: for building test build and for running tests using same (?) build.
Separately these cases work:
1) -xctestrun **without** build-for-testing (runs tests)
2) -scheme 'Bikemap' **with** build-for-testing and **without** -xctestrun (produces a new .xctestrun file)

**Comment by OlKir** [comment]
And it is strange but it seems that fix from [this branch](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/pull/144) doesn't work - I still see build-for-testing parameter there. 
I've double checked and I have proper test-center version:
`Using fastlane-plugin-test_center 3.8.4 from https://github.com/lyndsey-ferguson/fastlane-plugin-test_center.git (at issue-142-scheme-transitent-testing-crash@ae10975)`
Removing it manually from Scanfile fixed xcodebuild fail (error related to Scheme) for me.

**Comment by lyndsey-ferguson** [comment]
When you write:
> I still see build-for-testing parameter there.

Is that in the `xcargs` of the `scan` summary?
```
| xcargs                                         | -UseNewBuildSystem=YES build-for-testing                          |
```
?

Also, are you using a `Scanfile`? Is `build-for-testing` there?

**Comment by lyndsey-ferguson** [comment]
I can duplicate the problem *if* I add `build-for-testing` to `xcargs` and _if_ I turn off the code that strips it out before calling `scan` for testing.

**Comment by OlKir** [comment]
1. build-for-testing was present in both scan summary and between parameters in this command (I saw this output in console):
```
set -o pipefail && env NSUnbufferedIO=YES xcodebuild -destination 'platform=iOS Simulator,id=AD30C630-9288-44FB-B96B-99F283B6BF95' -derivedDataPath 'build/DerivedData' -disable-concurrent-testing -xctestrun 'build/DerivedData/Build/Products/Bikemap_iphonesimulator13.0-x86_64.xctestrun' -UseNewBuildSystem=YES build-for-testing -parallel-testing-enabled NO -only-testing:BikemapTests/BMQuadTreeTests/testLargeTreeNearestSearch -only-testing:BikemapTests/BMQuadTreeTests/testLargeTreeSetup -only-testing:BikemapTests/BMQuadTreeTests/testMTKDuplicates -only-testing:BikemapTests/BMDirectionsTest/testDirectionResponse -only-testing:BikemapTests/BMNavigationSETests/testBarcelona -only-testing:BikemapTests/BMNavigationSETests/testDegreeMod -only-testing:BikemapTests/BMNavigationSETests/testSaikogasse -only-testing:BikemapTests/BMNavigationSETests/testABCD test-without-building | tee '/Users/oleksii/Library/Logs/scan/Bikemap-Bikemap.log' | xcpretty  --report junit --output '/Data/Bikemap/bikemap-x-ios/fastlane/test_output/BikemapTests-batch-1/report.junit' --report junit --output '/var/folders/rj/69jcm5w12sn3288smsn20t880000gn/T/junit_report20190925-55774-1fe0h5l'
```
2. Yes, I use Scanfile and build-for-testing was added there in xcargs. After I removed it tests run again. build-for-testing is still present in scan summary but not in the command.

So for me code which should strip it out doesn't work, I had to remove build-for-testing from my Scanfile manually.

**Comment by lyndsey-ferguson** [comment]
When I add `build_for_testing` to a `Scanfile`, it shows up its own parameter in the `scan` summary. I'm looking at how to make it so that `scan` does not pick it up.

**Comment by lyndsey-ferguson** [comment]
Ok, I've updated the branch with code that should solve the issue. It will have a different SHA so your Gemfile.lock file will not reference the correct code. Can you remove the Gemfile.lock, and then [repeat the install instructions](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/issues/142#issuecomment-534677144)?

Please use the `--verbose` flag when calling fastlane.

**Comment by OlKir** [comment]
It was my fault - I installed your modified gem using bundler and then run fastlane normally so it used unmodified gem.. After I run fastlane using bundler it worked as you expected. So I can confirm that fix from your branch removes build_for_testing from xargs list and this prevents xcodebuild from crashing!
Thank you very much for your effort, and sorry for confusion by the first reports. We're looking forward to have changes from fix branch in production!

**Comment by lyndsey-ferguson** [comment]
Fixed in [v3.8.5](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/tree/v3.8.5)
